### DISCUSSION TITLE
SSH control master

### DISCUSSION THREAD

**Discussion by Twink-Polar** [kingToolbox/WindTerm#141] [General]
You mentioned next in the list of planned feature is ssh control master. Curiosity really, what are you hoping to do with it?

**Comment by kingToolbox** [comment]
Thank you for your concern. I think ControlMaster should be a good feature. ControlMaster that enables the sharing of multiple sessions over a single network connection. This means that we can connect to the server once, and have all other subsequent ssh sessions use the initial connection without need for re-authentication. This will significantly speed up the time to open new duplicated ssh shell and sftp, especially when the network quality is not good. 

In addition, taking this opportunity, I am also refactoring the network module of WindTerm. In the old design, the login wizard, SSH authentication, shell, and filer(sftp, local file system) were all implemented in the same c++ class, which made it impossible to open a shell or filer separately. So this time I put the above functions into modules independently, which greatly improves the freedom of use of these functions. Let me give three examples to illustrate the advantages of such a change. 

The first one, if you have used VSCode, you should use its [terminal](https://code.visualstudio.com/docs/editor/integrated-terminal). Now WindTerm/WindEdit have the same feature, a `shell` pane, and furthermore, you can open any configured session, which means you can interact with the servers more conveniently, especially when you edit files on the server in WindEdit. 

The second one, not only WindEdit can have the `shell` pane, but also the `filer` pane. When editing a file, `filer` will open sftp or local flle system in it, which is convenient for opening or managing other files quickly. 

The third one, WindTerm will have the ability to quickly transfer files between two servers. Usually, you will use `Scp` command or `Sftp` clients when you need to transfer files between servers. But neither one is very convenient. After refactoring, you can upload files from the server to another server just like uploading local files to the server. After clicking the menu item `Download from other server...`, `Upload to other server...` in the `filer`, a dialog box for downloading or uploading to other servers will open. You can select the server you want to download/upload, and enter the file system of that server, and select the files you want to download or the directory you want to upload, and WindTerm will do the rest for you. Although the direct transmission of the `Scp` command is more efficient. But it's also fast and more convenient, especially when transferring multiple files and folders between servers. 

The above mentioned are all related to ControlMaster. They are so convenient because there is no need to log in multiple times. Of course, automatic login can also have a similar effect, but first, it will increase the burden on the server, and second, the speed will be slower. 

Unfortunately, since the initial design did not consider the independent use of features, the changes are very large. Almost 800~1000 functions and more than10,000 lines of code need to be refactored. So the progress is a bit slow, and I am still trying to do this. The most important and most difficult thing is that all existing functions need to be ensured to be normal. This requires me to re-test almost all functions. But I think it is worth it.

**Reply by Twink-Polar** [reply]
Thank you for the detailed response. Sounds very exciting the changes you are making by implementing ControlMaster functionality. 

Looking forward to trying it out.

**Reply by kingToolbox** [reply]
If all goes well, the WindTerm_2.2.0_Prerelease_1 will be released at the beginning of next week at the latest. Welcome to try it out and give your valuable feedback. :smiley:
